src/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncContext.Unix.cs

```diff
@@ -370,126 +418,410 @@ private sealed class SendFileOperation : WriteOperation
+                            Debug.Assert(_tail == null, "State == Ready but queue is not empty!");
+                            _sequenceNumber++;
+                            if (TraceEnabled) Trace(context, $"Exit (previously ready)");
+                            return;
+
+                        case QueueState.Waiting:
+                            Debug.Assert(_tail != null, "State == Waiting but queue is empty!");
+                            _state = QueueState.Processing;
+                            // Break out and release lock
+                            break;
+
+                        case QueueState.Processing:
+                            Debug.Assert(_tail != null, "State == Processing but queue is empty!");
+                            _sequenceNumber++;
+                            if (TraceEnabled) Trace(context, $"Exit (currently processing)");
+                            return;
+
+                        case QueueState.Stopped:
+                            Debug.Assert(_tail == null);
+                            if (TraceEnabled) Trace(context, $"Exit (stopped)");
+                            return;
+
+                        default:
+                            Environment.FailFast("unexpected queue state");
+                            return;
+                    }
                 }
 
-                head.Next = null;
-                operation = (TOperation)head;
-                return true;
+                // We just transitioned from Waiting to Processing.
+                // Spawn a work item to do the actual processing.
+                ThreadPool.QueueUserWorkItem(s_processingCallback, context);
             }
 
-            private void Requeue(TOperation operation)
+            // Called on the threadpool when data may be available.
+            public void ProcessQueue(SocketAsyncContext context)
             {
-                // Insert at the head of the queue
-                Debug.Assert(!IsStopped, "Expected !IsStopped");
-                Debug.Assert(operation.Next == null, "Operation already in queue");
+                int observedSequenceNumber;
+                AsyncOperation op;
+                using (Lock())
+                {
+                    if (TraceEnabled) Trace(context, $"Enter");
+
+                    if (_state == QueueState.Stopped)
+                    {
+                        Debug.Assert(_tail == null);
+                        if (TraceEnabled) Trace(context, $"Exit (stopped)");
+                        return;
+                    }
+                    else
+                    {
+                        Debug.Assert(_state == QueueState.Processing, $"_state={_state} while processing queue!");
+                        Debug.Assert(_tail != null, "Unexpected empty queue while processing I/O");
+                        observedSequenceNumber = _sequenceNumber;
+                        op = _tail.Next;        // head of queue
```

Comment:
At most one ProcessQueue invocation can execute at a time.  This is necessary to preserve ordering semantics.

I'll figure out how to validate this better in debug.